src/viewport_renderer.rs

```diff
@@ -40,23 +44,110 @@ pub struct Vertex {
 // save memory
 /// The geometry indices as uploaded on the GPU.
-pub type Index = u32;
+pub type RendererIndex = u32;
 
-/// The geometry containing index and vertex data in the format as
+/// The geometry containing index and vertex data in flat format as
 /// will be uploaded on the GPU.
 #[derive(Debug, Clone, PartialEq)]
-pub struct Geometry {
-    indices: Option<Vec<Index>>,
-    vertex_data: Vec<Vertex>,
+pub struct RendererGeometry {
+    indices: Option<Vec<RendererIndex>>,
+    vertex_data: Vec<RendererVertex>,
 }
 
-impl Geometry {
+impl RendererGeometry {
+    /// Construct flat geometry with same-length per-vertex data from
+    /// variable length data `Geometry`
+    pub fn from_geometry(geometry: &Geometry) -> Self {
+        let vertices = geometry.vertices();
+        if let Some(normals) = geometry.normals() {
+            // Duplicate either vertices or normals, whichever list
+            // was shorter
+            let vertices_len = vertices.len();
+            let normals_len = normals.len();
+            let len = vertices_len.max(normals_len);
+
+            let indices_len = geometry.triangle_face_indices_len();
+            let mut indices: Vec<RendererIndex> = Vec::with_capacity(indices_len);
+            let mut vertex_positions: Vec<Point3<f32>> = vec![Point3::origin(); len];
+            let mut vertex_normals: Vec<Vector3<f32>> = vec![Vector3::zeros(); len];
+
+            match vertices_len.cmp(&normals_len) {
+                Ordering::Less => {
+                    // If number of vertices is smaller than the
+                    // number of normals, the order of normals will be
+                    // used and vertices will be duplicated.
+
+                    for triangle_index in geometry.triangle_face_indices() {
+                        let v = triangle_index.vertices;
+                        let n = triangle_index
+                            .normals
+                            .expect("Normal indices must be present if normals are");
+                        let (v1, v2, v3) = (cast_usize(v.0), cast_usize(v.1), cast_usize(v.2));
+                        let (n1, n2, n3) = (cast_usize(n.0), cast_usize(n.1), cast_usize(n.2));
+
+                        indices.push(n.0);
+                        indices.push(n.1);
+                        indices.push(n.2);
+
+                        // Swizzle vertices to use normal indexing scheme
+                        vertex_positions[n1] = vertices[v1];
+                        vertex_positions[n2] = vertices[v2];
+                        vertex_positions[n3] = vertices[v3];
+
+                        // Copy normal data as is
+                        vertex_normals[n1] = normals[n1];
+                        vertex_normals[n2] = normals[n2];
+                        vertex_normals[n3] = normals[n3];
+                    }
+                }
+                Ordering::Greater | Ordering::Equal => {
```

Comment:
I should test this with a simple triangle/quad with only one normal. The conversion should decompress the normal and replicate it across vertices.